From d4f941a053b07b988b79b9f75b2300cadaaf9940 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jonas=20=C3=85dahl?= <jadahl@gmail.com>
Date: Mon, 31 Mar 2025 15:02:23 +0200
Subject: [PATCH 5/5] drm-lease: Treat connectors as unleasable when inactive
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

When inactive (logind session no longer active, e.g. VT switched to
another user, no longer DRM master) mutter stays dormant and doesn't
try to reconfigure outputs on hotplug events, among other things. That
means we also shouldn't keep leases active, or connectors leasible, when
not active, since dependent state are not kept up to date.

Address this by, treat all connectors as not leasable when inactive,
which effectively means leases are revoked.

This fixes a crash when monitors are hot plugged in a certain way when
the session is inactive, which an added test case tests.

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4378>
(cherry picked from commit eec0f21e40e4319b24f16eb4f97c783b4f559a3e)
---
 src/backends/native/meta-drm-lease.c | 36 ++++++++++++++++++++++++++++
 src/tests/drm-lease-monitor-test.c   | 17 +++++++++++++
 2 files changed, 53 insertions(+)

diff --git a/src/backends/native/meta-drm-lease.c b/src/backends/native/meta-drm-lease.c
index c6bd2d7380..6f0c64dae5 100644
--- a/src/backends/native/meta-drm-lease.c
+++ b/src/backends/native/meta-drm-lease.c
@@ -15,6 +15,8 @@
  * along with this program; if not, see <http://www.gnu.org/licenses/>.
  */
 
+#include "backends/native/meta-backend-native-types.h"
+#include "backends/native/meta-backend-native.h"
 #include "config.h"
 
 #include "backends/native/meta-drm-lease.h"
@@ -71,6 +73,8 @@ struct _MetaDrmLeaseManager
   gulong resources_changed_handler_id;
   gulong lease_changed_handler_id;
   gulong monitors_changed_handler_id;
+  gulong backend_pause_handler_id;
+  gulong backend_resume_handler_id;
 
   /* MetaKmsDevice *kms_device */
   GList *devices;
@@ -84,6 +88,8 @@ struct _MetaDrmLeaseManager
    * Value: MetaDrmLease *lease
    */
   GHashTable *leased_connectors;
+
+  gboolean is_paused;
 };
 
 G_DEFINE_TYPE (MetaDrmLeaseManager, meta_drm_lease_manager, G_TYPE_OBJECT)
@@ -665,6 +671,9 @@ update_connectors (MetaDrmLeaseManager  *lease_manager,
   new_leased_connectors =
     g_hash_table_new_similar (lease_manager->leased_connectors);
 
+  if (lease_manager->is_paused)
+    goto scanned_resources;
+
   for (l = meta_kms_get_devices (kms); l; l = l->next)
     {
       MetaKmsDevice *kms_device = l->data;
@@ -697,6 +706,8 @@ update_connectors (MetaDrmLeaseManager  *lease_manager,
         }
     }
 
+scanned_resources:
+
   g_hash_table_iter_init (&iter, lease_manager->leased_connectors);
   while (g_hash_table_iter_next (&iter, (gpointer *)&kms_connector, NULL))
     {
@@ -862,6 +873,21 @@ on_lease_changed (MetaKms             *kms,
   update_leases (lease_manager);
 }
 
+static void
+on_pause (MetaBackendNative   *backend_native,
+          MetaDrmLeaseManager *lease_manager)
+{
+  lease_manager->is_paused = TRUE;
+  update_resources (lease_manager);
+}
+
+static void
+on_resume (MetaBackendNative   *backend_native,
+           MetaDrmLeaseManager *lease_manager)
+{
+  lease_manager->is_paused = FALSE;
+}
+
 static void
 meta_drm_lease_manager_constructed (GObject *object)
 {
@@ -870,6 +896,7 @@ meta_drm_lease_manager_constructed (GObject *object)
   MetaBackend *backend = meta_kms_get_backend (kms);
   MetaMonitorManager *monitor_manager =
     meta_backend_get_monitor_manager (backend);
+  MetaBackendNative *backend_native = META_BACKEND_NATIVE (backend);
 
   /* Connect to MetaKms::resources-changed using G_CONNECT_AFTER to make sure
    * MetaMonitorManager state is up to date. */
@@ -885,6 +912,10 @@ meta_drm_lease_manager_constructed (GObject *object)
     g_signal_connect_swapped (monitor_manager, "monitors-changed-internal",
                               G_CALLBACK (update_resources),
                               lease_manager);
+  lease_manager->backend_pause_handler_id =
+    g_signal_connect (backend_native, "pause", G_CALLBACK (on_pause), lease_manager);
+  lease_manager->backend_resume_handler_id =
+    g_signal_connect (backend_native, "resume", G_CALLBACK (on_resume), lease_manager);
 
   lease_manager->leases =
     g_hash_table_new_full (NULL, NULL,
@@ -939,11 +970,16 @@ meta_drm_lease_manager_dispose (GObject *object)
   MetaBackend *backend = meta_kms_get_backend (kms);
   MetaMonitorManager *monitor_manager =
     meta_backend_get_monitor_manager (backend);
+  MetaBackendNative *backend_native = META_BACKEND_NATIVE (backend);
 
   g_clear_signal_handler (&lease_manager->resources_changed_handler_id, kms);
   g_clear_signal_handler (&lease_manager->lease_changed_handler_id, kms);
   g_clear_signal_handler (&lease_manager->monitors_changed_handler_id,
                           monitor_manager);
+  g_clear_signal_handler (&lease_manager->backend_pause_handler_id,
+                          backend_native);
+  g_clear_signal_handler (&lease_manager->backend_resume_handler_id,
+                          backend_native);
 
   g_list_free_full (g_steal_pointer (&lease_manager->devices), g_object_unref);
   g_list_free_full (g_steal_pointer (&lease_manager->connectors),
diff --git a/src/tests/drm-lease-monitor-test.c b/src/tests/drm-lease-monitor-test.c
index d35f5ba83a..9451036c44 100644
--- a/src/tests/drm-lease-monitor-test.c
+++ b/src/tests/drm-lease-monitor-test.c
@@ -78,11 +78,28 @@ test_drm_lease_lease_suspend_resume (void)
   fake_udev_hotplug ();
 }
 
+static void
+test_drm_lease_lease_suspend_no_resume (void)
+{
+  MetaBackend *backend = meta_context_get_backend (test_context);
+  MetaBackendNative *backend_native = META_BACKEND_NATIVE (backend);
+
+  drm_mock_set_resource_filter (DRM_MOCK_CALL_FILTER_GET_CONNECTOR,
+                                disconnect_connector_filter, NULL);
+  fake_udev_hotplug ();
+  meta_backend_native_pause (backend_native);
+
+  drm_mock_unset_resource_filter (DRM_MOCK_CALL_FILTER_GET_CONNECTOR);
+  fake_udev_hotplug ();
+}
+
 static void
 init_tests (void)
 {
   g_test_add_func ("/wayland/drm-lease/suspend-resume",
                    test_drm_lease_lease_suspend_resume);
+  g_test_add_func ("/wayland/drm-lease/suspend-no-resume",
+                   test_drm_lease_lease_suspend_no_resume);
 }
 
 int
-- 
2.49.0

